src/main/java/ru/yandex/clickhouse/BalancedClickhouseDataSource.java

```diff
@@ -0,0 +1,234 @@
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.logging.Logger;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+import static ru.yandex.clickhouse.ClickhouseJdbcUrlParser.JDBC_CLICKHOUSE_PREFIX;
+
+public class BalancedClickhouseDataSource implements DataSource {
+    private static final org.slf4j.Logger log = LoggerFactory.getLogger(BalancedClickhouseDataSource.class);
+    private static final Pattern URL_TEMPLATE = Pattern.compile(JDBC_CLICKHOUSE_PREFIX + "//([a-zA-Z0-9_:,.]+)(/[a-zA-Z0-9_]+)?");
+
+    protected PrintWriter printWriter;
+    protected int loginTimeout = 0;
+    protected final ClickHouseDriver driver = new ClickHouseDriver();
+
+    private Random rnd = new Random();
+    private Set<String> disabledUrls = new HashSet<String>();
+    private Set<String> urls = new HashSet<String>();
+
+    private ClickHouseProperties properties;
+
+    public static List<String> splitUrl(final String url) {
+        Matcher m = URL_TEMPLATE.matcher(url);
+        if (!m.matches()) {
+            throw new IllegalArgumentException("Incorrect url");
+        }
+        String host = m.group(1);
+        String database = m.group(2);
+        if (database == null) {
+            database = "";
+        }
+        String[] hosts = host.split(",");
+        final List<String> res = new ArrayList<String>(hosts.length);
+        for (final String h : hosts) {
+            res.add(JDBC_CLICKHOUSE_PREFIX + "//" + h + database);
+        }
+        return res;
+    }
+
+
+    public BalancedClickhouseDataSource(final String url) {
+        this(splitUrl(url), new ClickHouseProperties());
+    }
+
+    public BalancedClickhouseDataSource(final String url, Properties info) {
+        this(splitUrl(url), new ClickHouseProperties(info));
+    }
+
+    public BalancedClickhouseDataSource(final String url, ClickHouseProperties properties) {
+        this(splitUrl(url), properties);
+    }
+
+    private BalancedClickhouseDataSource(final List<String> urls) {
+        this(urls, new ClickHouseProperties());
+    }
+
+    private BalancedClickhouseDataSource(final List<String> urls, Properties info) {
+        this(urls, new ClickHouseProperties(info));
+    }
+
+    private BalancedClickhouseDataSource(final List<String> urls, ClickHouseProperties properties) {
+        if (urls.isEmpty()) {
+            throw new IllegalArgumentException("Incorrect ClickHouse jdbc url list. It must be not empty");
+        }
+
+        try {
+            this.properties = ClickhouseJdbcUrlParser.parse(urls.get(0), properties.asProperties());
+            properties.setHost(null);
+            properties.setPort(-1);
+        } catch (URISyntaxException e) {
+            throw new IllegalArgumentException(e);
+        }
+
+        for (final String url : urls) {
+            addUrl(url);
+        }
+    }
+
+    private void disableUrl(final String url) {
+        synchronized (urls) {
```

Comment:
I use synchronization because there is no need lock on read, only on write. But, probably, you are right. ConcurrentHashMap seems to be a better solution.